storage/integration_test/src/integration_test.cc

```diff
@@ -630,6 +642,402 @@ TEST_F(FirebaseStorageTest, TestDeleteFile) {
                     firebase::storage::kErrorObjectNotFound);
 }
 
+// Only test retries on desktop since Android and iOS don't have an option
+// to retry file-not-found errors and just pass-through to native
+// implementations.
+#if FIREBASE_PLATFORM_DESKTOP
+TEST_F(FirebaseStorageTest, TestGetBytesWithMaxRetryDuration) {
+  // Enable retrying of file-not-found errors for testing.
+  bool old_value =
+      firebase::storage::internal::g_retry_file_not_found_for_testing;
+  firebase::storage::internal::g_retry_file_not_found_for_testing = true;
+
+  int shorter_duration_sec = 2;
+  int longer_duration_sec = 6;
+  SignIn();
+
+  // Call GetBytes on a non-existent ref. Call PutBytes while the GetBytes is
+  // still retrying. Verify that GetBytes succeeds.
+  {
+    LogDebug("Call PutBytes while GetBytes is retrying.");
+    firebase::storage::StorageReference ref = CreateFolder().Child("File2.txt");
+    LogDebug("Storage URL: gs://%s%s", ref.bucket().c_str(),
+             ref.full_path().c_str());
+    cleanup_files_.push_back(ref);
+
+    std::thread t1([&] {
+      // PutBytes after a short delay.
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(1000 * shorter_duration_sec));
+      LogDebug("Upload sample file from memory.");
+      firebase::Future<firebase::storage::Metadata> future =
+          RunWithRetry<firebase::storage::Metadata>([&]() {
+            return ref.PutBytes(&kSimpleTestFile[0], kSimpleTestFile.size());
+          });
+      WaitForCompletion(future, "PutBytes");
+      auto metadata = future.result();
+      EXPECT_EQ(metadata->size_bytes(), kSimpleTestFile.size());
+    });
+    // GetBytes with a long retry duration.
+    const size_t kBufferSize = 1024;
+    char buffer[kBufferSize];
+    memset(buffer, 0, sizeof(buffer));
+    storage_->set_max_download_retry_time(longer_duration_sec);
+    firebase::Future<size_t> future = ref.GetBytes(buffer, kBufferSize);
+    WaitForCompletion(future, "GetBytes");
+    ASSERT_NE(future.result(), nullptr);
+    size_t file_size = *future.result();
+    EXPECT_EQ(file_size, kSimpleTestFile.size());
+    EXPECT_THAT(kSimpleTestFile, ElementsAreArray(buffer, file_size))
+        << "Download failed, file contents did not match.";
+
+    t1.join();
+  }
+
+  // Call GetBytes on a non-existent ref. Call PutBytes after GetBytes should
+  // have stopped retrying. Verify that GetBytes fails.
+  {
+    LogDebug("Call PutBytes after the maximum retry deadline.");
+    firebase::storage::StorageReference ref = CreateFolder().Child("File3.txt");
+    LogDebug("Storage URL: gs://%s%s", ref.bucket().c_str(),
+             ref.full_path().c_str());
+    cleanup_files_.push_back(ref);
+
+    std::thread t1([&] {
+      // PutBytes after a long delay.
+      std::this_thread::sleep_for(
```

Comment:
Here you could (and should) actually check the status of the original future to ensure it's completed. You can even call WaitForCompletion in multiple threads at once.